lib/services/local/generic/generic.go

```diff
@@ -75,17 +79,22 @@ func (c *ServiceConfig[T]) CheckAndSetDefaults() error {
 		return trace.BadParameter("unmarshal func is missing")
 	}
 
+	if c.RunWhileLockedRetryInterval < 0 && !testing.Testing() {
```

Comment:
Do we want to allow a negative retry interval when testing? This addition of `if testing` makes this code hard to test, as it behaves differently while running tests and in production. IMO validation should not depend on who is running the code.

Reply:
Yes and we need this because 
https://github.com/gravitational/teleport/blob/b178b8b7322e7c6348bbd17ed1d3251ad34d946b/lib/backend/helpers.go#L107-L109

Will block when using the fake clock. It's really difficult to unblock because we need to call `clock.Advance` when it's locked in that condition otherwise we can't continue which is almost impossible to handle when multiple calls to RunWhileLocked happen in //.
Given that, passing a negative value automatically unblocks the retry call.

We already have some usecases of `testing.Testing` within Teleport to prevent some actions that might be really bad for users but aren't bad for  tests like changing the bcrypt cost flag,

Reply:
Using `testing.Testing` to speed up test execution is an entirely different scenario. Reducing the cost of the bcrypt function, you don't change the invariant of the code where, here, a different error will be returned when you run the same code with `go test` and `go run`.
From what I see, you use this code only in one place https://github.com/gravitational/teleport.e/pull/3857/files#diff-1f8bd6884fd83302647f81fcb6a2b86637b04f8fb48da58b949293fbf5d5529dR774
Probably, the most correct way to change this is to pass a function that returns a channel that can be overridden from a test (or change the `LockConfiguration` struct to an interface so a different implementation can be passed). 
The second solution, which requires fewer changes, is to allow negative values in test and normal passes with a comment that a negative value fires immediately. That at least makes the testing match the behavior in production.
Third, allow zero as a timeout but make the timer optional/pointer so that it is not set by default.

I'm pretty sure there are also some alternative solutions, and I'm happy to approve this PR as long as we don't use `testing.Testing()` to alter the behavior of tests.

Reply:
> Third, allow zero as a timeout but make the timer optional/pointer so that it is not set by default.

We set it by default to 250ms and teleport relies on it.

I removed usage from testing.Testing but I do think we should check it anw. It's trivial to add a patch that breaks a lot of stuff by passing invalid values.